### ISSUE TITLE
Duplicated join requests for the same user in the same space

### ISSUE THREAD

**Issue by mconf-daileon** [mconf/mconf-web#616]
---

Author Name: **Leonardo Daronco** (Leonardo Daronco)
Original Redmine Issue: 1278, http://dev.mconf.org/redmine/issues/1278

Original Assignee: Leonardo Daronco

---

A user should not be able to request to join a space if he already has a pending request, either if he was the one that asked to join or if he was invited to join.
Also, invitations should not be sent to users in the same situation (that already have a pending request).

In short, ":candidate_id" and ":space_id" should never be duplicated in JoinRequest. This can be enforced in the model only, but maybe it would be nice to also force it in the database.

**Comment by mconf-daileon** [comment]
---

Original Redmine Comment
Author Name: **Lucas Zawacki** (Lucas Zawacki)
Original Date: 2014-07-03T16:49:24Z

---

It's already here https://github.com/mconf/mconf-web/blob/branch-v2/app/models/join_request.rb#L28

Do we need anything more?

**Comment by mconf-daileon** [comment]
---

Original Redmine Comment
Author Name: **Leonardo Daronco** (Leonardo Daronco)
Original Date: 2014-07-03T17:43:36Z

---

It's better than I thought at first, but there are still a few things:

* Why are @:group_type@ and @:processed_at@ in the scope? Maybe @:group_type@ should not be there. If a user was already invited, he shouldn't be able to ask to join. If he already asked to join, he shouldn't be able to be invited.
* Why @:allow_nil@? Is a join request without a candidate useful? Being used in the events? For spaces there's always a candidate now, never just an email.
* The message when you try to invite someone and he already has an invitation is not very clear, maybe we can improve it.

**Comment by mconf-daileon** [comment]
---

Original Redmine Comment
Author Name: **Lucas Zawacki** (Lucas Zawacki)
Original Date: 2014-07-04T20:28:20Z

---

Almost everything is now addressed here https://github.com/mconf/mconf-web/pull/22

group_type is there because group is a polymorphic field and processed is there because we can have different join request for the same user if the past ones have been denied (the processed_at field will have different dates).
